metadata/metadata.go

```diff
@@ -111,14 +111,38 @@ func Join(mds ...MD) MD {
 type mdIncomingKey struct{}
 type mdOutgoingKey struct{}
 
-// NewIncomingContext creates a new context with incoming md attached.
+// NewIncomingContext creates a new context with incoming md attached. If used
+// in conjunction with AppendToOutgoingContext, NewIncomingContext will
+// overwrite any previously-appended metadata.
 func NewIncomingContext(ctx context.Context, md MD) context.Context {
-	return context.WithValue(ctx, mdIncomingKey{}, md)
+	return context.WithValue(ctx, mdIncomingKey{}, rawMD{md: md})
 }
 
 // NewOutgoingContext creates a new context with outgoing md attached.
 func NewOutgoingContext(ctx context.Context, md MD) context.Context {
-	return context.WithValue(ctx, mdOutgoingKey{}, md)
+	return context.WithValue(ctx, mdOutgoingKey{}, rawMD{md: md})
+};
+
+// AppendToOutgoingContext returns a new context with the provided kv merged
+// with any existing metadata in the context. Pairs should be provided
+// as k1,v1,k1,v2,k2,v3,...
```

Comment:
How about "Please refer to the documentation of Pairs for a description of kv"?

Reply:
Done